cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,241 @@
+ */
+template <bool find_first>
+auto __device__ element_index_pair_iter(size_type const size)
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
+  if constexpr (find_first) {
+    return thrust::pair(thrust::make_counting_iterator(0), thrust::make_counting_iterator(size));
   } else {
-    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
+    return thrust::pair(thrust::make_reverse_iterator(thrust::make_counting_iterator(size)),
+                        thrust::make_reverse_iterator(thrust::make_counting_iterator(0)));
   }
 }
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
-  }
+template <typename Type, typename SearchKeyIter>
+struct search_lists_fn {
+  cudf::detail::lists_column_device_view const lists;
+  SearchKeyIter const keys_iter;
+  duplicate_find_option const find_option;
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+  __device__ thrust::pair<size_type, bool> operator()(size_type const idx) const
+  {
+    auto const list = list_device_view{lists, idx};
+    // A null list never contains any key, even null key.
+    // In addition, a null list will result in a null output row.
+    if (list.is_null()) { return {NOT_FOUND_SENTINEL, false}; }
+
+    auto const key_opt = keys_iter[idx];
+    // A null key will also result in a null output row.
+    if (!key_opt) { return {NOT_FOUND_SENTINEL, false}; }
+
+    auto const& key = key_opt.value();
+    return {find_option == duplicate_find_option::FIND_FIRST ? search_list<true>(list, key)
+                                                             : search_list<false>(list, key),
+            true};
   }
-}
 
-/**
- * @brief __device__ functor to search for a key in a `list_device_view`.
- */
-template <duplicate_find_option find_option>
-struct finder {
-  template <typename ElementType>
-  __device__ size_type operator()(list_device_view const& list, ElementType const& search_key) const
+ private:
+  template <bool find_first>
+  static __device__ size_type search_list(list_device_view const& list, Type const& search_key)
   {
-    auto const list_begin = find_begin<ElementType, find_option>(list);
-    auto const list_end   = find_end<ElementType, find_option>(list);
-    auto const find_iter  = thrust::find_if(
-      thrust::seq, list_begin, list_end, [search_key] __device__(auto element_and_validity) {
-        auto [element, element_is_valid] = element_and_validity;
-        return element_is_valid && cudf::equality_compare(element, search_key);
+    auto const [begin, end] = element_index_pair_iter<find_first>(list.size());
```

Comment:
Is this function going to be used elsewhere? Otherwise I think it would be much easier to read as an IILE

Reply:
This will be reused in the follow-up PR when nested types implementation is added.